File: webflux-dialogue/src/test/java/com/study/webflux/rag/domain/dialogue/model/ConversationTurnTest.java
package com.study.webflux.rag.domain.dialogue.model;

import java.time.Instant;

import com.study.webflux.rag.fixture.ConversationSessionFixture;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class ConversationTurnTest {

	@Test
	@DisplayName("유효한 쿼리로 ConversationTurn 생성 성공")
	void create_validQuery_success() {
		ConversationSessionId sessionId = ConversationSessionFixture.createId();
		String query = "안녕하세요";

		ConversationTurn turn = ConversationTurn.create(sessionId, query);

		assertThat(turn.sessionId()).isEqualTo(sessionId);
		assertThat(turn.query()).isEqualTo(query);
		assertThat(turn.id()).isNull();
		assertThat(turn.response()).isNull();
		assertThat(turn.createdAt()).isNotNull();
	}

	@Test
	@DisplayName("null 쿼리로 생성 시 예외 발생")
	void create_nullQuery_throwsException() {
		ConversationSessionId sessionId = ConversationSessionFixture.createId();
		assertThatThrownBy(() -> ConversationTurn.create(sessionId, null))
			.isInstanceOf(IllegalArgumentException.class)
			.hasMessageContaining("query cannot be null or blank");
	}

	@Test
	@DisplayName("빈 문자열 쿼리로 생성 시 예외 발생")
	void create_blankQuery_throwsException() {
		ConversationSessionId sessionId = ConversationSessionFixture.createId();
		assertThatThrownBy(() -> ConversationTurn.create(sessionId, "   "))
			.isInstanceOf(IllegalArgumentException.class)
			.hasMessageContaining("query cannot be null or blank");
	}

	@Test
	@DisplayName("응답 추가")
	void withResponse_addsResponse() {
		ConversationSessionId sessionId = ConversationSessionFixture.createId();
		ConversationTurn turn = ConversationTurn.create(sessionId, "질문");
		String response = "답변입니다";

		ConversationTurn updatedTurn = turn.withResponse(response);

		assertThat(updatedTurn.response()).isEqualTo(response);
		assertThat(updatedTurn.query()).isEqualTo(turn.query());
		assertThat(updatedTurn.createdAt()).isEqualTo(turn.createdAt());
	}

	@Test
	@DisplayName("ID와 함께 ConversationTurn 생성")
	void withId_createsWithAllFields() {
		ConversationSessionId sessionId = ConversationSessionFixture.createId();
		String id = "test-id";
		String query = "질문";
		String response = "답변";
		Instant now = Instant.now();

		ConversationTurn turn = ConversationTurn.withId(id, sessionId, query, response, now);

		assertThat(turn.id()).isEqualTo(id);
		assertThat(turn.sessionId()).isEqualTo(sessionId);
		assertThat(turn.query()).isEqualTo(query);
		assertThat(turn.response()).isEqualTo(response);
		assertThat(turn.createdAt()).isEqualTo(now);
	}

	@Test
	@DisplayName("withId로 null 쿼리 생성 시 예외 발생")
	void withId_nullQuery_throwsException() {
		ConversationSessionId sessionId = ConversationSessionFixture.createId();
		assertThatThrownBy(
			() -> ConversationTurn.withId("id", sessionId, null, "response", Instant.now()))
			.isInstanceOf(IllegalArgumentException.class)
			.hasMessageContaining("query cannot be null or blank");
	}
}

File: app-miyou/src/test/kotlin/com/miyou/app/domain/dialogue/model/ConversationTurnTest.kt
package com.miyou.app.domain.dialogue.model

import com.miyou.app.fixture.ConversationSessionFixture
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.time.Instant

class ConversationTurnTest {
    @Test
    @DisplayName("유효한 질문으로 대화 턴을 생성한다")
    fun create_validQuery_success() {
        val sessionId = ConversationSessionFixture.createId()
        val query = "hello"

        val turn = ConversationTurn.create(sessionId, query)

        assertThat(turn.sessionId).isEqualTo(sessionId)
        assertThat(turn.query).isEqualTo(query)
        assertThat(turn.id).isNull()
        assertThat(turn.response).isNull()
        assertThat(turn.createdAt).isNotNull()
    }

    @Test
    @DisplayName("질문이 비어 있으면 대화 턴을 생성할 수 없다")
    fun create_blankQuery_throwsException() {
        val sessionId = ConversationSessionFixture.createId()

        assertThatThrownBy { ConversationTurn.create(sessionId, "   ") }
            .isInstanceOf(IllegalArgumentException::class.java)
            .hasMessageContaining("query cannot be null or blank")
    }

    @Test
    @DisplayName("응답을 포함한 복사본을 반환한다")
    fun withResponse_addsResponse() {
        val sessionId = ConversationSessionFixture.createId()
        val turn = ConversationTurn.create(sessionId, "question")
        val response = "answer"

        val updatedTurn = turn.withResponse(response)

        assertThat(updatedTurn.response).isEqualTo(response)
        assertThat(updatedTurn.query).isEqualTo(turn.query)
        assertThat(updatedTurn.createdAt).isEqualTo(turn.createdAt)
    }

    @Test
    @DisplayName("식별자를 포함한 모든 필드로 대화 턴을 생성한다")
    fun withId_createsWithAllFields() {
        val sessionId = ConversationSessionFixture.createId()
        val id = "test-id"
        val query = "question"
        val response = "answer"
        val now = Instant.now()

        val turn = ConversationTurn.withId(id, sessionId, query, response, now)

        assertThat(turn.id).isEqualTo(id)
        assertThat(turn.sessionId).isEqualTo(sessionId)
        assertThat(turn.query).isEqualTo(query)
        assertThat(turn.response).isEqualTo(response)
        assertThat(turn.createdAt).isEqualTo(now)
    }

    @Test
    @DisplayName("질문이 비어 있으면 식별자 기반 생성에 실패한다")
    fun withId_blankQuery_throwsException() {
        val sessionId = ConversationSessionFixture.createId()

        assertThatThrownBy {
            ConversationTurn.withId("id", sessionId, "   ", "response", Instant.now())
        }.isInstanceOf(IllegalArgumentException::class.java)
            .hasMessageContaining("query cannot be null or blank")
    }
}
